feat: Query if ScrollViewState is_at_bottom()

Teufelchen1 opened the pull request:
Hey 🐅,

this aims to fix #72 

It is close to what was discussed but I made the additional decision to respect the page size.
I believe this is the right way to do it as not respecting the page size feels very odd to the user, example:
```Rust
let state = ...;
// our content is 5 rows
assert_eq!(state.size.unwrap().height, 5);
// our page size is 3
assert_eq!(state.page_size.unwrap().height, 3);

// Scrolling down 5 times
state.scroll_down();
state.scroll_down();
state.scroll_down();
state.scroll_down();
state.scroll_down();

// Render it
scroll_view.render(...);

// Are we at the bottom?
assert!(state.is_at_bottom()); // FAILS!
// This is because rendering resets our offset to cater for the page size
assert_eq!(state.offset.unwrap().y, 2);
```
See `scroll_view.rs:210-232` for why that is.

Since that felt wrong to me, I made the judgment call for `is_at_bottom()` to consider the page size. So technically it is now a `are_we_scrolled_down_enough_so_it_behaves_and_looks_like_we_are_at_the_bottom()` 🥸

I added a unit test.

Feedback? :)

Teufelchen1 committed:
feat: Query if ScrollViewState is_at_bottom()

Teufelchen1 commented on tui-scrollview/src/scroll_view.rs:
```diff
@@ -451,6 +450,65 @@ mod tests {
+            ])
+        );
+
+        // We could also jump directly to the bottom
+        state.scroll_to_bottom();
+        assert!(state.is_at_bottom());
+
+        // which sets the offset to the last row of content,
+        // ensuring to be at the bottom regardless of the page size.
+        assert_eq!(state.offset.y, state.size.unwrap().height - 1);
```
More a kind of explanation & documentation rather than a strict unit test.

joshka commented on tui-scrollview/src/scroll_view.rs:
```diff
@@ -451,6 +450,65 @@ mod tests {
+        // => We see the top five rows
+        assert_eq!(state.offset.y, 0);
+        assert!(!state.is_at_bottom());
+        assert_eq!(
+            buf,
+            Buffer::with_lines(vec![
+                "ABCDE▲",
+                "KLMNO█",
+                "UVWXY█",
+                "EFGHI║",
+                "OPQRS▼",
+                "◄██═► ",
+            ])
+        );
```
I tend to prefer not putting these sort of pre-condition assertions that are covered elsewhere in tests. My rationale for this preference is that if some other change fails to render the starting condition, I'd prefer to see this fail at the point where it's testing the end condition, rather than in the pre-conditions. As a general rule, I've found that this makes it easier to debug failing tests while developing as well as to maintain them when things change which would affect pre-conditions but not post-conditions.

Teufelchen1 replied:
Yes, I am with you on that opinion. Dropped it.

joshka commented on tui-scrollview/src/scroll_view.rs:
```diff
@@ -787,10 +845,10 @@ mod tests {
         let mut buf = Buffer::empty(Rect::new(0, 0, 7, 5));
         let mut state = ScrollViewState::default();
         let mut list_state = ListState::default();
-        let items: Vec<String> = (1..=10).map(|i| format!("Item {}", i)).collect();
+        let items: Vec<String> = (1..10).map(|i| format!("Item {}", i)).collect();
```
This was intentional. I wanted 10 items, not 9 in the list. What's the purpose of this change?

Teufelchen1 replied:
Whoops, that wasn't supposed to be in the commit. Change dropped.

joshka commented on tui-scrollview/src/scroll_view.rs:
```diff
@@ -787,10 +845,10 @@ mod tests {
         let mut buf = Buffer::empty(Rect::new(0, 0, 7, 5));
         let mut state = ScrollViewState::default();
         let mut list_state = ListState::default();
-        let items: Vec<String> = (1..=10).map(|i| format!("Item {}", i)).collect();
+        let items: Vec<String> = (1..10).map(|i| format!("Item {}", i)).collect();
         let list = List::new(items);
         scroll_view.render_stateful_widget(list, scroll_view.area(), &mut list_state);
-        scroll_view.render(buf.area, &mut buf, &mut state);
+        scroll_view.clone().render(buf.area, &mut buf, &mut state);
```
I'm not sure why this clone is needed here

Teufelchen1 replied:
It is actually reassuring to see that you also didn't get it. I was very confused when Rust hit me with this (when you omit that `clone()`):
```txt
error[E0382]: use of moved value: `scroll_view`
   --> tui-scrollview/src/scroll_view.rs:482:9
    |
454 |     fn move_to_bottom(scroll_view: ScrollView) {
    |                       ----------- move occurs because `scroll_view` has type `scroll_view::ScrollView`, which does not implement the `Copy` trait
...
461 |         scroll_view.render(buf.area, &mut buf, &mut state);
    |                     -------------------------------------- `scroll_view` moved due to this method call
...
482 |         scroll_view.render(buf.area, &mut buf, &mut state);
    |         ^^^^^^^^^^^ value used here after move
    |
note: `ratatui::prelude::StatefulWidget::render` takes ownership of the receiver `self`, which moves `scroll_view`
   --> /home/teufelchen/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ratatui-0.29.0/src/widgets.rs:244:15
    |
244 |     fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State);
    |               ^^^^
help: you can `clone` the value and consume it, but this might not be your desired behavior
    |
461 |         scroll_view.clone().render(buf.area, &mut buf, &mut state);
    |                    ++++++++

For more information about this error, try `rustc --explain E0382`.

```

I have to admit I am unable to explain it really. I figured it has something to do with `#[fixture]` or `#[rstest]` which I don't know about. Hence I just went with the compiler hint / the clone.

joshka commented on tui-scrollview/src/state.rs:
```diff
@@ -82,4 +82,14 @@ impl ScrollViewState {
         self.offset.y = bottom;
     }
+
+    /// True if the scroll view state is at the bottom of the buffer
+    /// Takes the pagesize into account
```
nit: Wrap at 100 chars. First line should be a summary sentence followed by a newline then any detailed info.

Here, I think it's worth adding some info on the caveats such as if the size / page size is not yet computed (by rendering), then it won't be taken into account. Likely this info should be also added to scroll_to_bottom docs too.

Teufelchen1 replied:
I can't wrap at 100 chars as the line is not longer than 100 chars. Or my editor lies to me :shaking_face: 

I added some info on the behavior of both functions. While doing so, I figured it would be best to just define a known behavior in that case. Changing the default for the buffer from `u16:Max` to `0` ensures that the function always returns true in that case. Also spotted a potential panic/undefined behavior, if the scroll position is at `u16:Max` and pagesize is `>= 1`, the computation would overflow the underlying `u16`. Saturated add seems fine to me, objections?

Teufelchen1 replied:
> As it stands wrapping with a single line would not cause a new paragraph, so the second line is a continuation of the first. It's wrapped too soon.


Is it correct now?

joshka commented on tui-scrollview/src/state.rs:
```diff
@@ -82,4 +85,17 @@ impl ScrollViewState {
+    ///
+    /// This takes the page size into account. It returns true if the current scroll offset
+    /// plus the page size matches or exceeds the buffer length.
+    /// The buffer and the page size are unkown until computed during the first rendering.
```
Typo unkown

joshka committed:
test: characterize scrollview bottom edge cases

Document the current bottom-state problems before fixing them: bottom
detection reports true before the last row is visible, and rendering at
the bottom stores a page size clipped by remaining content instead of
the viewport height.

joshka committed:
fix: correct scrollview bottom detection

Treat the page end as exclusive when checking whether the last content
row is visible. Store the scrollbar-adjusted viewport size before
clipping to the backing buffer so page-based scrolling remains
consistent after rendering at the bottom.

joshka committed:
test: cover scrollview scrollbar visibility policy
